File: server/utils/aiService.js
const { GoogleGenerativeAI } = require('@google/generative-ai');

const compareAttempts = async (problemDescription, originalApproach, originalCode, newApproach, newCode) => {
    if (!process.env.GEMINI_API_KEY) {
        console.warn('GEMINI_API_KEY not found, returning mock response');
        return {
            summary: 'AI Feedback unavailable (API Key missing).',
            rating: 'REMEMBERED',
        };
    }

    const genAI = new GoogleGenerativeAI(process.env.GEMINI_API_KEY);
    const model = genAI.getGenerativeModel({ model: 'gemini-pro' });

    const prompt = `
    You are an expert coding tutor. Compare the user's current review attempt with their original solution and the problem description.
    
    Problem: ${problemDescription}
    
    Original Approach: ${originalApproach}
    Original Code: ${originalCode}
    
    Current Approach: ${newApproach}
    Current Code: ${newCode}
    
    Provide a JSON response with two fields:
    1. "summary": A brief feedback summary (max 3 sentences) on how well they remembered the solution and any improvements/regressions.
    2. "rating": One of ["REMEMBERED", "PARTIAL", "FORGOT"].
       - REMEMBERED: Logic is correct and matches optimal/original approach.
       - PARTIAL: Some details missing or suboptimal but core idea is there.
       - FORGOT: Completely wrong or stuck.
       
    Output ONLY valid JSON.
  `;

    try {
        const result = await model.generateContent(prompt);
        const response = await result.response;
        const text = response.text();

        // Clean up markdown code blocks if present
        const jsonStr = text.replace(/```json/g, '').replace(/```/g, '').trim();

        return JSON.parse(jsonStr);
    } catch (error) {
        console.error('AI Service Error:', error);
        return {
            summary: 'Error generating AI feedback.',
            rating: 'PARTIAL', // Default to partial on error to be safe
        };
    }
};

module.exports = { compareAttempts };

File: legacy_python/ai_service.py
import os
import google.generativeai as genai
import json
from typing import Optional

# Configure Gemini
GEMINI_API_KEY = os.environ.get("GEMINI_API_KEY")
if GEMINI_API_KEY:
    genai.configure(api_key=GEMINI_API_KEY)

def compare_attempts(problem_description: str,
                     original_approach: str,
                     original_code: Optional[str],
                     new_approach: str,
                     new_code: Optional[str]) -> dict:
    """
    Calls Gemini and returns a dict:
    {
      "summary": str,
      "rating": str  # one of "REMEMBERED", "PARTIAL", "FORGOT"
    }
    """
    if not GEMINI_API_KEY:
        return {
            "summary": "Gemini API key not configured.",
            "rating": "FORGOT"
        }

    model = genai.GenerativeModel('gemini-pro')

    prompt = f"""
    You are an expert coding tutor. Compare the user's new attempt to solve a problem with their original attempt.
    
    Problem Description:
    {problem_description}

    Original Approach:
    {original_approach}

    Original Code:
    {original_code if original_code else "N/A"}

    New Approach:
    {new_approach}

    New Code:
    {new_code if new_code else "N/A"}

    Analyze how well the user remembered the solution. 
    Provide a JSON response with two fields:
    1. "summary": A concise summary of the comparison and feedback (max 3 sentences).
    2. "rating": One of "REMEMBERED" (solved correctly/similarly), "PARTIAL" (remembered core idea but missed details), or "FORGOT" (completely different or wrong).

    JSON Response:
    """

    try:
        response = model.generate_content(prompt)
        text = response.text
        # Clean up potential markdown code blocks
        if text.startswith("```json"):
            text = text[7:]
        if text.endswith("```"):
            text = text[:-3]
        
        result = json.loads(text)
        
        # Validate rating
        if result.get("rating") not in ["REMEMBERED", "PARTIAL", "FORGOT"]:
            result["rating"] = "FORGOT"
            
        return result
    except Exception as e:
        print(f"Error calling Gemini: {e}")
        return {
            "summary": "Error analyzing attempt.",
            "rating": "FORGOT"
        }
